sdk/storage/Azure.Storage.Blobs/src/PartitionedUploader.cs

```diff
@@ -0,0 +1,249 @@
+        }
+
+        public Response<BlobContentInfo> Upload(
+            Stream content,
+            BlobHttpHeaders blobHttpHeaders,
+            IDictionary<string, string> metadata,
+            BlobRequestConditions conditions,
+            IProgress<long> progressHandler,
+            AccessTier? accessTier = default,
+            CancellationToken cancellationToken = default)
+        {
+            if (TryGetLength(content, out long length) && length < _singleBlockThreshold)
+            {
+                return _client.Upload(content, blobHttpHeaders, metadata, conditions, accessTier, progressHandler, cancellationToken);
+            }
+
+            return UploadInSequence(content, blobHttpHeaders, metadata, conditions, progressHandler, accessTier, cancellationToken);
+        }
+
+        private Response<BlobContentInfo> UploadInSequence(
+            Stream content,
+            BlobHttpHeaders blobHttpHeaders,
+            IDictionary<string, string> metadata,
+            BlobRequestConditions conditions,
+            IProgress<long> progressHandler,
+            AccessTier? accessTier,
+            CancellationToken cancellationToken)
+        {
+            List<string> blocks = new List<string>();
+
+            IAsyncEnumerator<StreamPartition> enumerator = GetBlocksAsync(content, async: false, cancellationToken).
+                GetAsyncEnumerator(cancellationToken);
+
+            while (enumerator.MoveNextAsync().EnsureCompleted())
+            {
+                using StreamPartition block = enumerator.Current;
+                string blockId = GenerateBlockId(block.AbsolutePosition);
+
+                _client.StageBlock(
+                    blockId,
+                    new MemoryStream(block.Bytes, 0, block.Length, writable: false),
+                    conditions: conditions,
+                    progressHandler: progressHandler,
+                    cancellationToken: cancellationToken);
+
+                blocks.Add(blockId);
+            }
+
+            return _client.CommitBlockList(blocks, blobHttpHeaders, metadata, conditions, accessTier, cancellationToken);
+        }
+
+        private async Task<Response<BlobContentInfo>> UploadInParallelAsync(
+            Stream content,
+            BlobHttpHeaders blobHttpHeaders,
+            IDictionary<string, string> metadata,
+            BlobRequestConditions blobAccessConditions,
+            IProgress<long> progressHandler,
+            AccessTier? accessTier,
+            CancellationToken cancellationToken)
+        {
+            List<string> blocks = new List<string>();
+            List<Task> runningTasks = new List<Task>();
+
+            await foreach (StreamPartition block in GetBlocksAsync(content, async: true, cancellationToken))
+            {
+                string blockId = GenerateBlockId(block.AbsolutePosition);
```

Comment:
Any reason we're using the absolute position instead of a counter?  I'm kind of on the fence.  On the one hand it's easier to see everything going out sequentially in your logs.  On the other hand you can correlate the block with the data in your file.  Feel free to leave it.

Reply:
I don't have a strong opinion here.